config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java

```diff
@@ -34,11 +35,72 @@
         this.bundleDocumentBuilder = bundleDocumentBuilder;
     }
 
-    public Element build(Bundle bundle, EntityBuilder.BundleType bundleType, Document document) {
+    public Map<String, Element> build(Bundle bundle, EntityBuilder.BundleType bundleType, Document document) {
         List<Entity> entities = new ArrayList<>();
         entityBuilders.forEach(builder -> entities.addAll(builder.build(bundle, bundleType, document)));
+        List<Entity> annotatedEntities = new ArrayList<>();
+        //TODO this should be for annotated entities(service, encass, policy)
+        entityBuilders.forEach(builder -> annotatedEntities.addAll(builder.build(bundle, EntityBuilder.BundleType.DEPLOYMENT, document).stream().filter(e -> e.getType().equals("POLICY")).collect(Collectors.toList())));
+
+        Map<String, Element> annotatedElements = new HashMap<>();
+        for (Entity annotatedEntity : annotatedEntities) {
+
+            List<Entity> entityList = getEntityDependencies(annotatedEntity, entities, bundle);
+            LOGGER.log(Level.WARNING, "Annotated entity dependencies" + entityList);
+            if (EntityBuilder.BundleType.DEPLOYMENT == bundleType) {
+                entityList.add(annotatedEntity);
+            }
+            annotatedElements.put(annotatedEntity.getName(), bundleDocumentBuilder.build(document, entityList));
+        }
+        return annotatedElements;
+
+    }
+
+    private void populatedDependentFolders(List<Entity> entityDependenciesList, List<Entity> entities, GatewayEntity annotatedItem) {
+
+        if (annotatedItem instanceof Folderable) {
+            Folder folder = ((Folderable) annotatedItem).getParentFolder();
+            while (folder != null) {
+                final String id = folder.getId();
+                Optional<Entity> optionalEntity = entities.stream().filter(e -> id.equals(e.getId())).findFirst();
+                if (optionalEntity.isPresent()) {
+                    entityDependenciesList.add(optionalEntity.get());
+                }
+                folder = folder.getParentFolder();
+            }
```

Comment:
Can we get the all the Folder Id first and then do 1 iteration on entries to find all dependencies. It will reduce the time complexity. Something like,
```
final Set<String> folderIds = new HashSet<>();
while (folder != null) {
    folderIds.add(folder.getId());
    folder = folder.getParentFolder();
}
entityDependenciesList.addAll(entities.parallelStream().filter(e -> folderIds.contains(e.getId())).collect(Collectors.toList()));
```